File: word_pattern.py
"""Using two hashmaps

Time complexity will be O(n) + O(n) = O(n)
Space complexity will be O(1)"""

class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        words = s.split()
        if len(pattern) != len(words):
            return False

        pattern_words_map, words_pattern_map = {}, {}

        for p, word in zip(pattern, words):
            if p not in pattern_words_map and word not in words_pattern_map:
                pattern_words_map[p] = word
                words_pattern_map[word] = p
            elif p in pattern_words_map:
                if not pattern_words_map[p] == word:
                    return False
            elif word in words_pattern_map:
                if not words_pattern_map[word] == p:
                    return False
        return True

File: WordPattern.cs
//Time Complexity:O(N)
//Space Complexity:O(1)
//Code solved in leetcode: Yes
//Approach in 3 lines:

// 1. Split the given string `s` into individual words and ensure it matches the length of the pattern.
// 2. Map each character in the pattern to its corresponding word in `s`, ensuring consistent mappings and no repetition of words.
// 3. If all mappings are valid, return true; otherwise, return false.

public class Solution {
    public bool WordPattern(string pattern, string s) {
        Dictionary<char, string> result = new Dictionary<char, string>();
        var words = s.Split(' ');

        if(pattern.Length != words.Length)
        {
            return false;
        }
        for(int i=0; i<pattern.Length; i++)
        {
            if(result.TryGetValue(pattern[i], out var currentword))
            {
                if(currentword != words[i])
                {
                     return false;
                }
                continue;
            }
            if(result.ContainsValue(words[i]))
            {
                return false;
            }
            result.Add(pattern[i],words[i]);
        }
        return true;
    }
}